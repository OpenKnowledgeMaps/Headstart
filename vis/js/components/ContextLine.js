import React from "react";
import { connect } from "react-redux";

import ContextLineTemplate from "../templates/ContextLine";
import Author from "../templates/contextfeatures/Author";
import DocumentTypes from "../templates/contextfeatures/DocumentTypes";
import NumArticles from "../templates/contextfeatures/NumArticles";
import DataSource from "../templates/contextfeatures/DataSource";
import Timespan from "../templates/contextfeatures/Timespan";
import PaperCount from "../templates/contextfeatures/PaperCount";
import DatasetCount from "../templates/contextfeatures/DatasetCount";
import Funder from "../templates/contextfeatures/Funder";
import ProjectRuntime from "../templates/contextfeatures/ProjectRuntime";
import LegacySearchLang from "../templates/contextfeatures/LegacySearchLang";
import SearchLang from "../templates/contextfeatures/SearchLang";
import Timestamp from "../templates/contextfeatures/Timestamp";
import Modifier from "../templates/contextfeatures/Modifier";
import MoreInfoLink from "../templates/contextfeatures/MoreInfoLink";
import MetadataQuality from "../templates/contextfeatures/MetadataQuality";
import ContextTimeFrame from "../templates/contextfeatures/ContextTimeFrame";
import DocumentLang from "../templates/contextfeatures/DocumentLang";

const defined = (param) => param !== undefined && param !== null;

/**
 * Component that creates the heading contextline.
 *
 * It has to be a class component because of the popovers (they use 'this').
 */
class ContextLine extends React.Component {
  render() {
    const { params, localization, hidden, service } = this.props;
    const { popoverContainer } = this.props;

    if (hidden) {
      return null;
    }

    return (
      <ContextLineTemplate>
          {params.showAuthor && (
              <Author
                  bioLabel={localization.bio_link}
                  livingDates={params.author.livingDates}
                  link={"https://d-nb.info/gnd/" + params.author.id}
              />
          )}
          <NumArticles
              articlesCount={params.articlesCount}
              openAccessArticlesCount={params.openAccessCount}
              articlesCountLabel={localization.articles_label}
          >
              <Modifier popoverContainer={popoverContainer}/>
          </NumArticles>
          {defined(params.dataSource) && (
              <DataSource
                  label={localization.source_label}
                  source={params.dataSource}
                  contentProvider={params.contentProvider}
                  popoverContainer={this.props.popoverContainer}
              />
          )}
          {defined(params.timespan) &&
              <Timespan>
                  <ContextTimeFrame popoverContainer={popoverContainer} timespan={params.timespan}/>
              </Timespan>}
          <DocumentTypes
              documentTypes={params.documentTypes}
              popoverContainer={popoverContainer}
          />
          {/* was an issue to left "All Languages" as default value in the context if no lang_id in parameters */}
          <DocumentLang
              value={params.documentLang}
              popoverContainer={popoverContainer}
          />
          {defined(params.paperCount) && (
              <PaperCount
                  value={params.paperCount}
                  label={localization.paper_count_label}
              />
          )}
          {defined(params.datasetCount) && (
              <DatasetCount
                  value={params.datasetCount}
                  label={localization.dataset_count_label}
              />
          )}
          {defined(params.funder) && <Funder>{params.funder}</Funder>}
          {defined(params.projectRuntime) && (
              <ProjectRuntime>{params.projectRuntime}</ProjectRuntime>
          )}
          {defined(params.legacySearchLanguage) && (
              <LegacySearchLang>{params.legacySearchLanguage}</LegacySearchLang>
          )}
          {defined(params.timestamp) && (
              <Timestamp
                  value={params.timestamp}
                  label={localization.timestamp_label}
              />
          )}
          <MetadataQuality
              quality={params.metadataQuality}
              service={service}
              popoverContainer={popoverContainer}
          />
          {defined(params.searchLanguage) && (
              <SearchLang>{params.searchLanguage}</SearchLang>
          )}
          <MoreInfoLink/>
      </ContextLineTemplate>
    );
  }
}

const mapStateToProps = (state) => ({
  hidden: state.zoom || !state.contextLine.show,
  params: {
    ...state.contextLine,
    timespan: state.timespan,
  },
  service: state.service,
  localization: state.localization,
});

export default connect(mapStateToProps)(ContextLine);
