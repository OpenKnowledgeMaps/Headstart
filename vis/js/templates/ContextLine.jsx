import React from "react";

const ContextLine = ({
  articlesCount, // context.num_documents
  articlesCountLabel, // config.localization[config.language].articles_label
  openAccessArticlesCount = null, // context.share_oa if config.show_context_oa_number
  dataSource = null, // config.service_name || config.service_names[context.service]
  dataSourceLabel, // config.localization[config.language].source_label
  timespan = null, // string composed from config data
  docTypes = null, // React component
  paperCount = null, // context.num_papers
  paperCountLabel, // config.localization[config.language].paper_count_label
  datasetCount = null, // context.num_datasets
  datasetCountLabel, // config.localization[config.language].dataset_count_label
  funder = null, // context.params.funder
  projectRuntime = null, // string composed from config data
  searchLang = null, // string composed from multiple params in config.options.languages
  timestamp = null, // context.last_update
  timestampLabel, // config.localization[config.language].timestamp_label
  modifier = null, // React component
  author = null, // React component
}) => {
  return (
    // html template starts here
    <p id="context">
      {author}
      <span id="num_articles">
        {articlesCount} {modifier}
        {articlesCountLabel}{" "}
        {openAccessArticlesCount !== null && (
          <>({openAccessArticlesCount} open access)</>
        )}
      </span>
      {dataSource !== null && <span id="source">{dataSourceLabel}: {dataSource}</span>}
      {timespan !== null && <span id="timespan">{timespan}</span>}
      {docTypes}
      {paperCount !== null && (
        <span id="context-paper_count">
          {paperCount} {paperCountLabel}
        </span>
      )}
      {datasetCount !== null && (
        <span id="context-dataset_count">
          {datasetCount} {datasetCountLabel}
        </span>
      )}
      {funder !== null && <span id="context-funder">Funder: {funder}</span>}
      {projectRuntime !== null && (
        <span id="context-project_runtime">{projectRuntime}</span>
      )}
      {searchLang !== null && <span id="search_lang">{searchLang}</span>}
      {timestamp !== null && <span id="timestamp">{timestampLabel}: {timestamp}</span>}
    </p>
    // html template ends here
  );
};

export default ContextLine;

export const DocumentTypes = ({ label, popoverLabel, text, popover }) => {
  if (popover) {
    return (
      <span
        id="document_types"
        data-content={
          <>
            {popoverLabel}
            <br />
            <br />
            {text}
          </>
        }
        data-toggle="popover"
        data-trigger="hover"
        data-html="true"
        className="context_moreinfo"
        data-original-title=""
        title=""
      >
        {label}
      </span>
    );
  }

  return (
    <span id="document_types">
      {label}: {text}
    </span>
  );
};
