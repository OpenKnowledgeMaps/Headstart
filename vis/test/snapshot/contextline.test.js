import React from "react";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import { Provider } from "react-redux";

import ContextLine from "../../js/components/ContextLine";

const mockStore = configureStore([]);

const setup = (
  overrideContextLine = {},
  overrideLocalization = {},
  overrideStoreObject = {}
) => {
  const storeObject = Object.assign(
    {
      zoom: false,
      contextLine: {
        show: true,
        articlesCount: 100, // context.num_documents
        modifier: null, // context.params.sorting => most-recent || most-relevant
        showModifierPopover: false, // context.params.sorting === "most-relevant" && config.context_most_relevant_tooltip
        openAccessCount: null, // config.show_context_oa_number => context.share_oa
        showAuthor: false, // config.is_authorview && context.params.author_id
        author: {
          id: null, // context.params.author_id
          livingDates: null, // context.params.living_dates
        },
        documentTypes: null,
        dataSource: "Sample data source",
        timespan: null,
        paperCount: null,
        datasetCount: null,
        funder: null,
        projectRuntime: null,
        legacySearchLanguage: null,
        timestamp: null,
        ...overrideContextLine,
      },
      localization: {
        most_recent_label: "most recent",
        most_relevant_label: "most relevant",
        most_relevant_tooltip: "Sample most relevant tooltip",
        articles_label: "documents",
        bio_link: "Biografie",
        documenttypes_label: "Document types",
        documenttypes_tooltip: "Sample document types tooltip",
        source_label: "Source",
        paper_count_label: "papers",
        dataset_count_label: "datasets",
        timestamp_label: "Last updated",
        high_metadata_quality: "high metadata quality",
        high_metadata_quality_desc_base:
          "This knowledge map only includes documents with an abstract (min. 300 characters). High metadata quality significantly improves the quality of your knowledge map.",
        high_metadata_quality_desc_pubmed:
          "This knowledge map only includes documents with an abstract. High metadata quality significantly improves the quality of your knowledge map.",
        low_metadata_quality: "low metadata quality",
        low_metadata_quality_desc_base:
          "This knowledge map includes documents with and without an abstract. Low metadata quality may significantly reduce the quality of your knowledge map. ",
        low_metadata_quality_desc_pubmed:
          "This knowledge map includes documents with and without an abstract. Low metadata quality may significantly reduce the quality of your knowledge map. ",

        ...overrideLocalization,
      },
    },
    overrideStoreObject
  );

  return storeObject;
};

describe("Context line component snapshot", () => {
  it("matches a project website (base low quality) snapshot", () => {
    const store = mockStore(
      setup(
        {
          modifier: "most-relevant",
          showModifierPopover: true,
          openAccessCount: 52,
          documentTypes: ["Journal/newspaper article"],
          dataSource: "BASE",
          timespan: "All time",
          metadataQuality: "low",
        },
        {},
        {
          service: "base",
        }
      )
    );

    const tree = renderer
      .create(
        <Provider store={store}>
          <ContextLine />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("matches a project website (pubmed high quality) snapshot", () => {
    const store = mockStore(
      setup(
        {
          modifier: "most-relevant",
          showModifierPopover: true,
          openAccessCount: 52,
          documentTypes: ["Journal/newspaper article"],
          dataSource: "PubMed",
          timespan: "All time",
          metadataQuality: "high",
        },
        {},
        {
          service: "pubmed",
        }
      )
    );

    const tree = renderer
      .create(
        <Provider store={store}>
          <ContextLine />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("matches a covis snapshot", () => {
    const store = mockStore(
      setup(
        {
          articlesCount: 76,
          openAccessCount: 34,
          dataSource:
            '<span class="backlink"><a href="data" class="underline" target="_blank" >CoVis database</a></span>',
          timestamp: "2020-08-12 02:40:04 UTC",
        },
        {
          articles_label: "resources and collections",
          source_label: "Data source",
        }
      )
    );

    const tree = renderer
      .create(
        <Provider store={store}>
          <ContextLine />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("matches a viper snapshot", () => {
    const store = mockStore(
      setup({
        articlesCount: 36,
        openAccessCount: 10,
        dataSource: "OpenAIRE",
        paperCount: 36,
        datasetCount: 0,
        funder: "EC",
        projectRuntime: "2011-2016",
      })
    );

    const tree = renderer
      .create(
        <Provider store={store}>
          <ContextLine />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("matches a linkedcat authorview snapshot", () => {
    const store = mockStore(
      setup(
        {
          showAuthor: true,
          author: {
            id: "100016642",
            livingDates: "1791-1863",
          },
          articlesCount: 45,
          dataSource: "LinkedCat+",
        },
        {
          articles_label: "open access Dokumente",
          source_label: "Quelle",
        }
      )
    );

    const tree = renderer
      .create(
        <Provider store={store}>
          <ContextLine />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("matches a linkedcat keywordview snapshot", () => {
    const store = mockStore(
      setup(
        {
          articlesCount: 295,
          dataSource: "LinkedCat+",
          timespan: "1 Jan 1847 - 1 Jan 1918",
          documentTypes: [
            "Anthologie",
            "Bericht",
            "Biografie",
            "Briefsammlung",
          ],
        },
        {
          articles_label: "open access Dokumente",
          source_label: "Quelle",
          documenttypes_tooltip:
            "Die folgenden Publikationsarten wurden bei der Erstellung dieser Visualisierung in Betracht gezogen (nicht alle davon scheinen notwendigerweise in dieser Visualisierung auch auf)",
          documenttypes_label: "Dokumentarten",
        }
      )
    );

    const tree = renderer
      .create(
        <Provider store={store}>
          <ContextLine />
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
