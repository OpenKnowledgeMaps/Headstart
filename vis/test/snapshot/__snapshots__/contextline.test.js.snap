// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Context line component snapshot matches as covis snapshot 1`] = `
<p
  id="context"
>
  <span
    id="num_articles"
  >
    76
     
    <span
      className="modifier"
      id="modifier"
    >
      
    </span>
    resources and collections
     
    (
    34
     open access)
  </span>
  <span
    id="source"
  >
    Data source
    : 
    CoVis database
  </span>
  <span
    id="timestamp"
  >
    Last updated
    : 
    2020-08-12 02:40:04 UTC
  </span>
</p>
`;

exports[`Context line component snapshot matches as project website snapshot 1`] = `
<p
  id="context"
>
  <span
    id="num_articles"
  >
    100
     
    <span
      className="modifier context_moreinfo"
      data-content="Sample most relevant tooltip"
      data-original-title=""
      data-toggle="popover"
      data-trigger="hover"
      id="modifier"
      title=""
    >
      most relevant
       
    </span>
    documents
     
    (
    52
     open access)
  </span>
  <span
    id="source"
  >
    Source
    : 
    BASE
  </span>
  <span
    id="timespan"
  >
    All time
  </span>
  <span
    id="document_types"
  >
    Document types
    : 
    Journal/newspaper article
  </span>
</p>
`;
